edn/src/matcher.rs

```diff
@@ -72,6 +71,37 @@ impl<'a> Matcher<'a> {
         matcher.match_internal::<T>(value, pattern)
     }
 
+    // This is the same as itertools::diff_with() except that it:
+    // - returns a boolean
+    // - takes account of '...' (':ellipsis' for now)
+    // - automatically calls match_internal rather than requiring an is_equal() closure
+    fn match_iter_internal<I, T>(&self, v: I, p: I) -> bool
+        where I: IntoIterator<Item=&'a Value>,
+              T: PatternMatchingRules<'a, Value> {
+        let mut v = v.into_iter();
+        let mut p = p.into_iter();
+        while let Some(v_elem) = v.next() {
+            match p.next() {
+                None => return false,
+                Some(p_elem) => {
+                    match *p_elem {
+                        Value::Keyword(ref keyword) => {
+                            // We don't support having anything after the elipsis (yet?)
+                            match p.next() { Some(_) => unimplemented!(), None => {} }
+
+                            if keyword.0 == "ellipsis" { return true }
+                        },
+                        _ => {},
+                    }
+                    if !self.match_internal::<T>(&v_elem, &p_elem) {
+                        return false;
+                    }
+                },
+            }
+        }
+        match p.next() { None => true, Some(_) => false }
```

Comment:
So if I'm reading this correctly , the code does the following:
* iterate the "value" and "pattern" collections in lockstep, with the "value" collection being the driver
* if the "pattern" collection is finished while the "value" collection still has some elements to iterate over, bail with `false`.
* do a bunch of special casing, equality checking and peeking when encountering something considered to be an "ellipsis" as the currently iterated element in the "pattern" collection.
* recurse matching otherwise.
* if the "value" collection is finished while the "pattern" collection still has some elements to iterate over, bail with `false`.

This approach seems overly complex for a number of reasons. For example: `[1 2 3 ...] !~ [1 2 3]`, when it should. Similarly, `[1 ... 4] != [1 2 3 4]`. Unordered collections are even more interesting: `#{3 2 ... 4}` should match `#{1 2 3 4 5}`, and since symbols have a higher precedence than uints for example, I don't see how any of these can be achieved with lockstep iteration (in this case *everything* will come after the ellipsis).

Let's step back a bit and think about we're actually trying to achieve: surjection[1] between the two groups, where the "value" collection is the domain and the "pattern" collection is the codomain: all elements in the "value" collection have a correspondent in the "pattern" collection, and vice versa but with the special rule that `...` can match anything. You could interpret `...` as sort of an reverse `_`: make sure each item in the "value" collection has a correspondent in the "pattern" collection, where "correspondents" can be something that pass `match_internal` or are equal to `...`.

I expect that it should be possible to implement this with just easy calls to `Iterator` consumers [2], like it is for the `set` and `map` collections.

[1] https://en.wikipedia.org/wiki/Surjective_function
[2] https://doc.rust-lang.org/stable/book/iterators.html#consumers